File: iOS/Sources/AppReveal/State/StateBridge.swift
// Bridges app state, navigation, and feature flags to AppReveal

import Foundation

#if DEBUG

/// Conform your app state container to expose state snapshots.
public protocol StateProviding: AnyObject {
    func snapshot() -> [String: AnyCodable]
}

@MainActor
final class StateBridge {

    static let shared = StateBridge()

    private weak var stateProvider: StateProviding?
    private weak var navigationProvider: NavigationProviding?
    private weak var featureFlagProvider: FeatureFlagProviding?
    private var resetHandlers: [() -> Void] = []

    private init() {}

    func registerStateProvider(_ provider: StateProviding) {
        stateProvider = provider
    }

    func registerNavigationProvider(_ provider: NavigationProviding) {
        navigationProvider = provider
    }

    func registerFeatureFlagProvider(_ provider: FeatureFlagProviding) {
        featureFlagProvider = provider
    }

    /// Register a handler called when `reset_app_state` is invoked.
    func registerResetHandler(_ handler: @escaping () -> Void) {
        resetHandlers.append(handler)
    }

    // MARK: - Queries

    func getState() -> [String: AnyCodable] {
        stateProvider?.snapshot() ?? [:]
    }

    func getNavigationStack() -> [String: AnyCodable] {
        guard let nav = navigationProvider else { return [:] }
        return [
            "currentRoute": AnyCodable(nav.currentRoute),
            "navigationStack": AnyCodable(nav.navigationStack),
            "presentedModals": AnyCodable(nav.presentedModals)
        ]
    }

    func getFeatureFlags() -> [String: AnyCodable] {
        featureFlagProvider?.allFlags() ?? [:]
    }

    func resetState() {
        for handler in resetHandlers {
            handler()
        }
    }
}

#endif

File: Flutter/appreveal/lib/src/state/state_bridge.dart
// Provider interfaces and registry for app state, navigation, and feature flags.

/// Implement to expose app state to AppReveal's [get_state] tool.
abstract interface class StateProviding {
  Map<String, dynamic> snapshot();
}

/// Implement to expose navigation state to AppReveal's [get_navigation_stack] tool.
abstract interface class NavigationProviding {
  String get currentRoute;
  List<String> get navigationStack;
  List<String> get presentedModals;
}

/// Implement to expose feature flags to AppReveal's [get_feature_flags] tool.
abstract interface class FeatureFlagProviding {
  Map<String, dynamic> allFlags();
}

class StateBridge {
  static final shared = StateBridge._();
  StateBridge._();

  StateProviding? _stateProvider;
  NavigationProviding? _navigationProvider;
  FeatureFlagProviding? _featureFlagProvider;

  void registerStateProvider(StateProviding provider) {
    _stateProvider = provider;
  }

  void registerNavigationProvider(NavigationProviding provider) {
    _navigationProvider = provider;
  }

  void registerFeatureFlagProvider(FeatureFlagProviding provider) {
    _featureFlagProvider = provider;
  }

  Map<String, dynamic> getState() {
    final provider = _stateProvider;
    if (provider == null) return {'error': 'No StateProviding registered'};
    return provider.snapshot();
  }

  Map<String, dynamic> getNavigationStack() {
    final provider = _navigationProvider;
    if (provider == null) return {'error': 'No NavigationProviding registered'};
    return {
      'currentRoute': provider.currentRoute,
      'navigationStack': provider.navigationStack,
      'presentedModals': provider.presentedModals,
    };
  }

  Map<String, dynamic> getFeatureFlags() {
    final provider = _featureFlagProvider;
    if (provider == null) return {'error': 'No FeatureFlagProviding registered'};
    return provider.allFlags();
  }
}
